ls1intum/Artemis
# [Bugfix] Exam programming exercise - navigation status - on reload

### Checklist

### Motivation and Context


This PR is a bugfix to issue number 11 from https://confluence.ase.in.tum.de/pages/viewpage.action?spaceKey=ArTEMiS&title=Issues+Student+Subteam.

### Description

The issue is that upon reloading the exam, the programming exercises that are out of sync are still marked as synced. This is because out of sync property is stored only on the client and is initialized with `true` by default. Now in order to set the correct value to the `isSynced` property I fetch the commitState of all programming exercises and set the submission to be out of sync if there are any uncommitted changes. 

### Steps for Testing
1. Join an exam
2. Submit on a programming exercise
3. Make changes and wait 30 seconds _(auto save time for the code editor)_
4. Reload the page
  4.1. Re-join the exam
  4.2. Exam navigation bar and exercises overview table **should be** the same as before reloading
5. Open the exam in the new tab
  5.1. Re-join the exam
  5.2. Exam navigation bar and exercises overview table **should be** the same as before reloading.

**NOTE: Please double check in the developer network view that the additional request (GET: repositoryService.getStatus()) is only made on exam reload and is not made on initial session (initial exam start)**

### Demo
https://user-images.githubusercontent.com/51077603/126080998-6b743e5b-1498-4018-8991-439968368f3b.mov

### Special Thanks
Thank you @imstreetspirit and @FlorianGlombik for collaborating with me on this one 👯
